### PR TITLE
Scala Effect API

### PR DESCRIPTION
**Pull request by nb-ceffa**
This PR has the ambition to add effect friendly metric API for simple usage with different frameworks.

Usage example can be found at `scala-effect-api/src/test/scala/com/avast/metrics/examples/EffectMonitor.scala`

### CODE REVIEW

File: scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/Monitor.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+  def wrap[F[_]: Sync](monitor: SMonitor): F[Monitor[F]] = Sync[F].delay(wrapUnsafe(monitor))
+  def wrap[F[_]: Sync](monitor: SMonitor, naming: Naming): F[Monitor[F]] = Sync[F].delay(wrapUnsafe(monitor, naming))
+
+  def applyUnsafe[F[_]: Sync](monitor: JMonitor): Monitor[F] = applyUnsafe(monitor, Naming.defaultNaming())
+  def applyUnsafe[F[_]: Sync](monitor: JMonitor, naming: Naming): Monitor[F] = wrapUnsafe(SMonitor(monitor), naming)
+
+  def wrapUnsafe[F[_]: Sync](monitor: SMonitor): Monitor[F] = wrapUnsafe(monitor, Naming.defaultNaming())
+  def wrapUnsafe[F[_]: Sync](monitor: SMonitor, naming: Naming): Monitor[F] = new impl.MonitorImpl(monitor, naming)
+
```
Comment (tomasherman) [inline scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/Monitor.scala:37]: i would change these names to `wrap` and `wrapJava` ... i think the `apply`/`wrap` is confusing and doesn't really communicate much

File: scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/Timer.scala [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      def time[A](block: F[A])(implicit monad: Monad[F]): F[A] = {
+        for {
+          context <- timer.start
+          result <- block
+          _ <- timer.stop(context)
+        } yield result
+      }
+    }
+  }
+}
```
Comment (tomasherman) [inline scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/Timer.scala:27]: does this need to be in companion object? I'd suggest it doesn't add much being there and makes it more difficult to use - i would just slap it into the interface itself (same for the TimerPair below)

File: scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/TimerPair.scala [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+object TimerPair {
+  object Dsl {
+    implicit class TimerPairOps[F[_]](val timerPair: TimerPair[F]) extends AnyVal {
+      def time[T](action: F[T])(implicit sync: Sync[F]): F[T] = {
+        Resource
+          .makeCase(timerPair.start) {
+            case (ctx, ExitCase.Completed) => timerPair.stop(ctx).as(())
+            case (ctx, _) => timerPair.stopFailure(ctx).as(())
+          }
+      }.use(_ => action)
```
Comment (tomasherman) [inline scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/TimerPair.scala:28]: again, i'd move this to timerpair itself

File: scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/impl/GaugeFactoryImpl.scala [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import cats.effect.Sync
+import com.avast.metrics.scalaapi.{Monitor => SMonitor}
+import com.avast.metrics.scalaeffectapi.{Gauge, GaugeFactory}
+
+import java.util.concurrent.atomic.{AtomicLong, AtomicReference}
+
+private class GaugeFactoryImpl[F[_]: Sync](monitor: SMonitor) extends GaugeFactory[F] {
+  override def long(name: String): Gauge[F, Long] = new Gauge[F, Long] {
+    private[this] val valueRef = new AtomicLong()
+    private[this] val gauge = monitor.gauge(name)(() => valueRef.get())
```
Comment (tomasherman) [inline scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/impl/GaugeFactoryImpl.scala:12]: have you considered using [Ref](https://typelevel.org/cats-effect/docs/2.x/concurrency/ref) directly?

File: scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/impl/TimerPairImpl.scala [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+private class TimerPairImpl[F[_]: Sync](success: Timer[F], failure: Timer[F]) extends TimerPair[F] {
+  override def start: F[TimerPairContext] = {
+    for {
+      succCtx <- success.start
+      failCtx <- failure.start
+    } yield new TimerPairContext {
+      override def stop: F[Duration] = success.stop(succCtx)
+
+      override def stopFailure: F[Duration] = failure.stop(failCtx)
```
Comment (tomasherman) [inline scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/impl/TimerPairImpl.scala:18]: i think it's simpler to introduce a case class for this but up to you

File: build.sbt [hunk 6]
```diff
@@ -74,6 +75,18 @@ lazy val scalaApi = (project in file("scala-api"))
   )
   .dependsOn(api, jmx % "test")
 
+lazy val scalaEffectApi = (project in file("scala-effect-api"))
+  .settings(
+    commonSettings,
+    scalaSettings,
+    scalacOptions += "-language:higherKinds",
+    name := "metrics-scala-effect",
```
Comment (tomasherman) [inline build.sbt:83]: i would call it `...-cats-effect2`, or `-ce2` as we will probably need a `ce3` or `-cats-effect3` in the future

**Review by tomasherman — COMMENTED** [review]
looks good to me so far, just a couple of minor points

File: scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/impl/GaugeFactoryImpl.scala (hunk 4 above)
Comment (nb-ceffa) [reply scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/impl/GaugeFactoryImpl.scala:12]: good point

File: scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/Monitor.scala (hunk 1 above)
Comment (nb-ceffa) [reply scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/Monitor.scala:37]: ok, Il' rename it

File: scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/Timer.scala (hunk 2 above)
Comment (nb-ceffa) [reply scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/Timer.scala:27]: After my last change (generic context removel), there is no need for that. I'll move it.

File: scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/TimerPair.scala (hunk 3 above)
Comment (nb-ceffa) [reply scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/TimerPair.scala:28]: I'll move it.

File: scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/impl/TimerPairImpl.scala (hunk 5 above)
Comment (nb-ceffa) [reply scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/impl/TimerPairImpl.scala:18]: yes, you can introduce a case class that contains the contexts and then call the stop method with the corresponding context but this looks better to me

File: scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/TimerPair.scala [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+trait TimerPair[F[_]] {
+  trait TimerPairContext {
+    def stop: F[Duration]
+    def stopFailure: F[Duration]
+  }
+
+  def start: F[TimerPairContext]
+  def stop(context: TimerPairContext): F[Duration]
+  def stopFailure(context: TimerPairContext): F[Duration]
```
Comment (jendakol) [inline scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/TimerPair.scala:13]: This is inconsistent with the other monitors, e.g.
```java
    TimeContext start();
    void update(Duration duration);
    void updateFailure(Duration duration);
```
The `stop` methods are already in the `TimerPairContext`.

File: scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/Timer.scala [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+import java.time.Duration
+
+trait Timer[F[_]] {
+  type Context
+  def start: F[Context]
+  def stop(context: Context): F[Duration]
+}
+
+object Timer {
+  object Dsl {
```
[the remark this answers is not in the document]
Comment (nb-ceffa) [reply scala-effect-api/src/main/scala/com/avast/metrics/scalaeffectapi/Timer.scala]: already removed

**Review by jendakol — APPROVED** [review]
Nice job!

File: README.md [hunk 9]
```diff
@@ -10,8 +10,9 @@ Library for application monitoring. It's abstraction of metrics inspired by [Dro
 
 Main advantages of this library:
 1) Universal abstraction with misc. implementations
-1) Support of multiple exports at once (`MultiMonitor`)
-1) Scala API
+2) Support of multiple exports at once (`MultiMonitor`)
+3) Scala API
+4) Cats-Effect API
```
Comment (m-handy) [inline README.md:15]: Just to be consistent - do you want to call it `Cats-Effect` or `Scala Effect` ?

**Pull request merged**